net/single_port_listener_factory.go

```diff
@@ -0,0 +1,187 @@
+package net
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+type singlePortTLSListenerFactory struct {
+	listener net.Listener
+
+	forward map[string]*singlePortTLSListener
+	fLock   sync.Mutex
+	logger  *logrus.Entry
+
+	stopC chan struct{}
+}
+
+type SinglePortTLSListenerFactoryArgs struct {
+	TLSConfig *tls.Config
+	Address   string
+	Logger    *logrus.Logger
+}
+
+func NewSinglePortTLSListenerFactory(args *SinglePortTLSListenerFactoryArgs) (ListenerFactory, error) {
```

Comment:
I'm pondering if `SinglePortTLSListenerFactory` is a good name. It's true that the outcome is that each listener you create "share" a port, but actually it's 1 listener that is shared. So maybe `SharedTLSPortListnerFactory`?

Reply:
Nice! Something seemed off with the naming and couldn't place it. Labeling it Shared also generalizes it such that if another project wanted to use it, would be less ambiguous